src/Open3D/GUI/Events.h

```diff
@@ -0,0 +1,184 @@
+// ----------------------------------------------------------------------------
+// -                        Open3D: www.open3d.org                            -
+// ----------------------------------------------------------------------------
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 www.open3d.org
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <cstdint>
+
+namespace open3d {
+namespace gui {
+
+enum class MouseButton {
+    NONE = 0,
+    LEFT = (1 << 0),
+    MIDDLE = (1 << 1),
+    RIGHT = (1 << 2),
+    BUTTON4 = (1 << 3),
+    BUTTON5 = (1 << 4)
+};
+
+enum class KeyModifier {
+    NONE = 0,
+    SHIFT = (1 << 0),
+    CTRL = (1 << 1),  // win/linux: ctrl, macOS: command
+    ALT = (1 << 2),   // win/linux: alt, macOS: ctrl
+    META = (1 << 3),  // win/linux: windows key, macOS: option
+};
+
+struct MouseEvent {
+    enum Type { MOVE, BUTTON_DOWN, DRAG, BUTTON_UP, WHEEL };
+    Type type;
+    int x;
+    int y;
+    int modifiers;  // KeyModifiers ORed together
+    union {
+        struct {
+            int buttons;  // MouseButtons ORed together
+        } move;           // includes drag
+        struct {
+            MouseButton button;
+        } button;
+        struct {
+            int dx;
+            int dy;
+            bool isTrackpad;
+        } wheel;
+    };
+};
+
+struct TickEvent {};
+
+enum KeyName {
```

Comment:
I think the reason that it isn't is the question of how to handle the numbers, and to lesser extent the letters. Obviously `0` won't work (KeyName::0), although `A` (KeyName::A) would compile although it could be awkward in code. We could solve the numbers problem by using KEY_0 or something, but then KeyName::KEY_0 seems a bit redundant. Do you have any suggestions?